pkg/controller/servicecidrs/servicecidrs_controller.go

```diff
@@ -0,0 +1,641 @@
+					if prefixContainIP(k, ipPrefix.Addr()) {
+						cidrs.Insert(v.UnsortedList()...)
+					}
+					return false
+				})
+				if cidrs.Len() == 1 && cidrs.Has(cidr.Name) {
+					return false, nil
+				}
+				logger.V(4).Info("No IPv6 addresses referencing", "ServiceCIDR", cidr.String())
+			}
+		}
+	}
+	// There are no IPAddresses that depend on the existing ServiceCIDR, so
+	// it is safe to delete, remove finalizer.
+	logger.Info("ServiceCIDR no longer have orphan IPs", "ServiceCDIR", cidr.String())
+	return true, nil
+}
+
+func (c *Controller) addServiceCIDRFinalizerIfNeeded(ctx context.Context, cidr *networkingapiv1alpha1.ServiceCIDR) error {
+	for _, f := range cidr.GetFinalizers() {
+		if f == ServiceCIDRProtectionFinalizer {
+			return nil
+		}
+	}
+
+	patch := map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"finalizers": []string{ServiceCIDRProtectionFinalizer},
+		},
+	}
+	patchBytes, err := json.Marshal(patch)
+	if err != nil {
+		return err
+	}
+	_, err = c.client.NetworkingV1alpha1().ServiceCIDRs().Patch(ctx, cidr.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	klog.FromContext(ctx).V(4).Info("Added protection finalizer to ServiceCIDR", "ServiceCIDR", cidr.Name)
+	return nil
+
+}
+
+func (c *Controller) removeServiceCIDRFinalizerIfNeeded(ctx context.Context, cidr *networkingapiv1alpha1.ServiceCIDR) error {
+	found := false
+	for _, f := range cidr.GetFinalizers() {
+		if f == ServiceCIDRProtectionFinalizer {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil
+	}
+	patch := map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"$deleteFromPrimitiveList/finalizers": []string{ServiceCIDRProtectionFinalizer},
+		},
+	}
+	patchBytes, err := json.Marshal(patch)
+	if err != nil {
+		return err
+	}
+	_, err = c.client.NetworkingV1alpha1().ServiceCIDRs().Patch(ctx, cidr.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	klog.FromContext(ctx).V(4).Info("Removed protection finalizer to ServiceCIDs", "ServiceCIDR", cidr.Name)
+	return nil
+}
+
+func ipAddressToPrefix(ip *networkingapiv1alpha1.IPAddress) (netip.Prefix, error) {
+	labelFamily, ok := ip.Labels[networkingapiv1alpha1.LabelIPAddressFamily]
+	if !ok {
+		return netip.Prefix{}, fmt.Errorf("IPAddress without required label: %s", networkingapiv1alpha1.LabelIPAddressFamily)
+	}
+	mask := "/32"
+	if labelFamily == string(v1.IPv6Protocol) {
+		mask = "/128"
+	}
+	return netip.ParsePrefix(ip.Name + mask)
+}
+
+// prefixContainIP returns true if the given IP is contained with the prefix
+// but is not the network address and also, if IPv4, is not the broadcast address.
+// This is required because the Kubernetes allocators reserve these addresses
+// so IPAddresses can not block deletion of this ranges.
+func prefixContainIP(prefix netip.Prefix, ip netip.Addr) bool {
```

Comment:
As noted above the problem still exists in other places. Basically _everywhere_ you look up an IP in an iptree.

Reply:
I moved it as a method of the tree, does not make sense to replicate everywhere are you pointed out